### ISSUE TITLE
registerFactoryParam does not allow subclasses for param types?

### ISSUE THREAD

**Issue by nlfiedler** [fluttercommunity/get_it#62]
It looks like `registerFactoryParam()` does not allow invocation of the factory function in which one of the parameters is an object that is a subclass of the defined parameter type. Is my understanding correct, and is this expected behavior? Should I be using some other method instead?

For instance, I register my factory like so:

```
  getIt.registerFactoryParam<PackStoreDetails, PackStore, void>(
    (param1, param2) => buildStoreDetails(param1, param2),
  );
```

`PackStore` is a simple class that is extended by `PackStoreModel`, and an instance of the latter is passed to the factory method, at which point I get this error:

```
Assertion failed:
file:///Users/nfiedler/flutter/.pub-cache/hosted/pub.dartlang.org/get_it-4.0.1/lib/get_it_impl.dart:108:17
param1 == null || param1.runtimeType == param1Type
"Incompatible Type passed as param1\nexpected: PackStore actual: PackStoreModel"
```

Thanks

**Comment by jacovdbosch** [comment]
Got the same issue while using https://pub.dev/packages/freezed.

**Comment by escamoteur** [comment]
I probably should use `is` instead `==`. I will try if that helps

**Comment by nlangh** [comment]
Same with https://pub.dev/packages/built_value:
```
Failed assertion: 
line 108 pos 17: 'param1 == null || param1.runtimeType == param1Type': 
Incompatible Type passed as param1
E/flutter ( 2589): expected: Profile actual: _$Profile
```

**Comment by damascenaluiz** [comment]
I have the same problem, maybe the suggestion above should be included in a next version

**Comment by votruk** [comment]
@escamoteur replacing `==` with `is` works as I describe in https://github.com/fluttercommunity/get_it/issues/67

**Comment by escamoteur** [comment]
@votruk I tried it and it didn't work because `is` does not work between a runtime type and a variable of type `Type`.

if it works for you I really would like to know how.

**Comment by escamoteur** [comment]
Next version has this test removed because of limitation of Dart

**Comment by bDev10** [comment]
@escamoteur It looks like for asynchronous factories this issue still exists in newest 5.0.4 version. I got it when using registerFactoryParamAsync.

**Comment by escamoteur** [comment]
could you add a small test in a PR?

**Comment by escamoteur** [comment]
Sorry for the late reply

**Comment by escamoteur** [comment]
Forget it, I found the problem

**Comment by escamoteur** [comment]
Fixed in V5.0.6
